### ISSUE TITLE
Proposed simplifications and feature removals for the HNC API

### ISSUE THREAD

**Issue by adrianludwin** [kubernetes-sigs/hierarchical-namespaces#199]
We have a few features that I believe are rarely used but are hard to implement and maintain in HNC:

- Cascading deletion: the ability to remove an entire subtree of subnamespaces (that is, namespaces with `SubnamespaceAnchor` objects in their parents) in one call to the server.
- Safely deleting the HNC CRDs: since deleting an anchor deletes its associated namespace, deleting the anchor _CRD_ would, if implemented naively, delete all subnamespaces and their contents. To prevent this, HNC tries _not_ to delete a subnamespace if the anchor's only being deleted because the CRD is being deleted, but it's hard to tell if this is foolproof.

In addition, users have never really gotten used to the distinction between "full" and "sub" namespaces, and refer to all child namespaces as "subnamespaces."

So here's a potential plan to simplify the HNC API:

- By default, deleting an anchor will no longer delete its subnamespace. We will add a `deleteSubnamespace` flag to the anchor that, if set, will delete the associated subnamespace if the anchor is deleted _for any reason_, including the CRD being deleted. But it will only be intended to be used by the kubectl plugin in a new `kubectl hns delete` command, and will not be recursive.
- The `hnc.x-k8s.io/subnamespaceOf` annotation on the namespace will become the source-of-truth for _all_ hierarchies. All child namespaces will now be referred to as "subnamespaces," and it will no longer be an error to have a subns without an anchor. Subnses _with_ anchors will be called "anchored subnamespaces," and their only purpose (other than creating the subns) will be to delete the subns (if `deleteSubnamespace` is true) or to modify its managed labels and annotations.
- The `HierarchyConfiguration` object will still exist, but its `.spec.parent` and `.spec.allowCascadingDeletion` fields will be deprecated and eventually removed. In v1.1, the `.spec.parent` field will be used to _set_ the `subnamespaceOf` annotation if it doesn't exist, but otherwise it will be ignored. This means its only purposes will be to set managed labels and annotations (if there's no anchor, especially for root namespaces) and to report the status of the namespace.
- We may add a `kubectl hns delete-tree` command to replace server-side cascading deletion. But this code will be far less dangerous IMO, especially since we can do things like providing warnings and listing the namespaces that are about to be deleted.

Other than the change in deletion behaviour, these changes will be backwards compatible for existing objects (e.g. HNC itself will set the `subnamespaceOf` annotation if it's unset), and mostly backwards compatible for anyone using the `kubectl-hns` plugin. For example, `kubectl hns set --parent` will still work exactly as it did before, except that instead of modifying the `HierarchyConfiguration` object, it will modify an annotation on the namespace itself.

This points to the one breaking change I can think of (although I haven't thought through it _too_ hard yet): **anyone modifying the hierarchy structure must also have permission to edit the namespace definition itself, instead of the `HierarchyConfiguration` object. ** I'd _imagine_ that this is largely not a problem since it will often be the cluster admins who are changing the hierarchy, and they probably have namespace editing permissions anyway.

In summary, making these changes would:

- Better align HNC terminology with the way it's actually being used (e.g. "subnamespace")
- Improve safety and reduce complexity by not deleting subnamespaces with their anchors or allowing server-side cascading deletion.

Any thoughts on this?

**Comment by adrianludwin** [comment]
/cc @rjbez17 
/cc @erikgb 

wdyt about these proposal?

**Comment by erikgb** [comment]
@adrianludwin In general I am positive to the suggested changes, but it seems like there might be a few corner cases that would need special attention (some examples in the following, but not limited to):
- How should HNC react to a user trying to create a new `SubnamespaceAnchor`, but the namespace already exists (with or without the `subnamespaceOf` annotation)? The safest thing to do, would probably be to reject the request, but I suspect users might ask for more flexibility.
- Do you think the `subnamespaceOf` annotation should be deleted if orphaning a subnamespace by deleting the anchor?

Instead of your suggested `deleteSubnamespace` flag (`true`/`false`), I suggest a new enum field. Enums are a lot more flexible IMO and usually easier to understand the semantics for. In this case I would name it `deletionPolicy`, with valid values `Delete` and `Orphan`.

Why was the special handling for cascade deletion introduced in the first place? Doesn't a `SubnamespaceAnchor` control its subnamespace? If the achor is deleted, I would expect that to trigger a deletion of the controlled subnamespace. And if that subnamespace about to be deleted contains anchors, it will just continue until the tree underneath the initial anchor is deleted.

I fully support removing the attempt to guard subnamespaces from being deleted when deleting CRDs. Even if it could be a disastrous destructive operation, any cluster-admin should know to take special care when deleting (or replacing) CRDs.

**Comment by adrianludwin** [comment]
Today, you can create an anchor for an existing namespace as long as the namespace's annotation points to the parent (this is how you resolve a `MissingAnchor` condition). We _could_ add a restriction that you can only create the anchor if you have edit permissions on the child namespace, but I'd rather not.

No, I don't think the `subnamespaceOf` annotation should be deleted if the anchor is deleted. The namespace simply becomes unanchored at that point, which is always allowed.

Re `deletionPolicy` - sgtm. The default value would by `Orphan` or perhaps `Unanchor` (since the subns isn't actually orphaned just because the anchor is deleted).

Cascading deletion would actually be the default behaviour today, but because it's so dangerous (you could mean to delete one namespace and accidentally delete a whole tree), we added a webhook to stop you from doing it accidentally. Cascading deletion actually does nothing but disable that webhook for a subtree.

**Comment by rjbez17** [comment]
I'm not neccessarily against these proposed changes, but I've used both of these features widely in the past. 

For cascading deletion, we had a tree in our dev clusters like Sandboxes -> {dev_team} -> {developer sandbox x}.

From there, the developers would create subnamespaces. We had a cost saving policy that dev sandboxes were auto-deleted monthly, allowing them to recreate it needed the next day. We leveraged the cascading deletion for this because it was simple and we diddn't have to care how many subnamespaces the developers created.

Secondly, we have heavily used the `.spec.parent` field of `HierarchyConfiguration` for migrating older clusters. Dev teams didn't have permissions to create/edit namespaces, but had legacy namespaces "given" to them long ago. We used `HierarchyConfiguration.spec.parent` as a way for them to lazily opt-in to the newer hierarchy. They were in charge of adding it themselves as we only had one root and HNC was used more for given clusterwide info to all namespaces (self signed CA). 

Neither are huge deals, especially #2 as it is a bit of an anti-pattern, but they were really nice for us. 

So I'm curious if it is worthwhile to solicit feedback via a poll to see how widely it is used?

**Comment by adrianludwin** [comment]
Hmm. For cascading deletion, would the client-side solution help? E.g. `kubectl hns rmtree <namespace>`? Although what I really think is that it would be nice to change the _default_ since this would let us drop the webhook and most of the finalizers (especially if combined with #200). For example, we could have three `deletionPolicy` values - `Unanchor`, `DeleteChild`, and `DeleteTree`, where `DeleteTree` essentially looks through all anchors in descendant namespaces and changes them to `DeleteChild` before deleting the namespace.

Regarding `.spec.parent` - that's interesting. Perhaps we can leave them as alternatives? E.g.:

- If `subnamespaceOf` exists, it's used and overwrites anything in `.spec.parent` (I _think_ this is what happens today)
- If `subnamespaceOf` is missing, but `.spec.parent` exists, then HNC sets the `subnamespaceOf` field

That preserves your usecase but still allows us to align the "subnamespace" terminology.

**Comment by rjbez17** [comment]
Now that I think more on it, I think I prefer removing the option of setting it in the config spec. As it stands, it could be a privileged escalation as anyone could really set any parent, meaning they could get some secrets or something they shouldn't.

I also like the thoughts on the cascading delete but also am ok leaving it for client side.

**Comment by adrianludwin** [comment]
We already have webhooks to prevent you from setting a parent that you
don't also own, for exactly that reason. That wouldn't go away :)
See:
https://github.com/kubernetes-sigs/hierarchical-namespaces/blob/d3b59688b3ca175f228e427db772d673b997ccc8/internal/hierarchyconfig/validator.go#L312
